src/priceClient/priceClient.ts

```diff
@@ -0,0 +1,212 @@
+import assert from "assert";
+import { CoinGeckoPrice } from "../coingecko";
+import { Logger } from "../relayFeeCalculator"; // @todo: Relocate Logger to utils?
+
+export { Logger } from "../relayFeeCalculator"; // Permit adapters to import a common definition.
+export type TokenPrice = CoinGeckoPrice; // Temporary inversion; CoinGecko should source from here.
+
+// Represents valid source for spot prices (Across API, CoinGecko, ...)
+export interface PriceFeedAdapter {
+  readonly name: string;
+  getTokenPrice(address: string, currency: string, platform: string): Promise<TokenPrice>;
+  getTokenPrices(addresses: string[], currency: string, platform: string): Promise<TokenPrice[]>;
+}
+
+export type PriceCache = {
+  [address: string]: TokenPrice;
+};
+
+export function msToS(ms: number): number {
+  return Math.floor(ms / 1000);
+}
+
+export class PriceClient implements PriceFeedAdapter {
+  private static instance: PriceClient | undefined;
+  public readonly name: string = "PriceClient";
+  private _maxPriceAge = 300; // seconds
+  private priceFeeds: PriceFeedAdapter[] = [];
+  private prices: {
+    [platform: string]: {
+      [currency: string]: PriceCache;
+    };
+  } = {};
+
+  protected constructor(protected logger: Logger) {}
+
+  public static get(logger: Logger): PriceClient {
+    if (!this.instance) this.instance = new PriceClient(logger);
+
+    return this.instance;
+  }
+
+  get maxPriceAge(): number {
+    return this._maxPriceAge;
+  }
+
+  set maxPriceAge(age: number) {
+    assert(age >= 0);
+    this.logger.debug({
+      at: "PriceClient#maxPriceAge",
+      message: `Setting maxPriceAge (S) ${this._maxPriceAge} => ${age}.`,
+    });
+    this._maxPriceAge = age;
+  }
+
+  // @todo: Could be nice to specify a priority rather than relying on order of addition.
+  addPriceFeed(priceFeed: PriceFeedAdapter): void {
+    assert(!this.priceFeeds.includes(priceFeed), `Price feed ${priceFeed.name} already registered.`);
+    this.priceFeeds.push(priceFeed);
+    this.logger.debug({
+      at: "PriceClient#addPriceFeed",
+      message: `Appended new price feed: ${priceFeed.name}.`,
+    });
+  }
+
+  listPriceFeeds(): string[] {
+    return this.priceFeeds.map((priceFeed: PriceFeedAdapter) => priceFeed.name);
+  }
+
+  clearPriceFeeds(): void {
+    const cleared: string[] = this.priceFeeds.map((priceFeed: PriceFeedAdapter) => priceFeed.name);
+    this.priceFeeds = [];
+    this.logger.debug({
+      at: "PriceClient#clearPriceFeeds",
+      message: "Cleared all price feeds.",
+      cleared,
+    });
+  }
+
+  async getTokenPrice(address: string, currency = "usd", platform = "ethereum"): Promise<TokenPrice> {
```

Comment:
Missing a toBN() conversion somewhere in here; this will incorrectly return a floating-point price. Will add a comment to the tests to catch that as well.

Reply:
Followed up on this point - the toBN() conversion occurs in the relayer ProfitClient, and it expected to receive a floating point price from CoinGecko. It should be OK to proceed with the current logic, as long as there is consensus on this implementation handling floating-point prices. Each adapter may eventually need to implement its own normalisation on the price format, though.